mdal/frmts/mdal_ugrid.cpp

```diff
@@ -401,3 +403,133 @@ void MDAL::DriverUgrid::parse2VariablesFromAttribute( const std::string &name, c
+  mNcFile.putAttrStr( mesh2dNodeXId, "long_name", "x-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "location", "node" );
+
+  // Nodes Y coordinate
+  int mesh2dNodeYId = mNcFile.defineVar( "mesh2d_node_y", NC_DOUBLE, 1, &dimNodeCountId );
+  mNcFile.putAttrStr( mesh2dNodeYId, "units", "m" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "standard_name", "projection_y_coordinate" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "long_name", "y-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "location", "node" );
+
+  // Nodes Z coordinate
+  int mesh2dNodeZId = mNcFile.defineVar( "mesh2d_node_z", NC_DOUBLE, 1, &dimNodeCountId );
+  mNcFile.putAttrStr( mesh2dNodeZId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "location", "node" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "coordinates", "mesh2d_node_x mesh2d_node_y" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "standard_name", "altitude" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "long_name", "z-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "units", "m" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "grid_mapping", "projected_coordinate_system" );
+  mNcFile.putAttrDouble( mesh2dNodeZId, "_FillValue", -999.0 );
+
+  // Faces 2D Variable
+  int mesh2FaceNodesId_dimIds [] { dimFaceCountId, dimMaxNodesPerFaceId };
+  int mesh2FaceNodesId = mNcFile.defineVar( "mesh2d_face_nodes", NC_INT, 2, mesh2FaceNodesId_dimIds );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "cf_role", "face_node_connectivity" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "location", "face" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "long_name", "Mapping from every face to its corner nodes (counterclockwise)" );
+  mNcFile.putAttrInt( mesh2FaceNodesId, "start_index", 1 );
+  mNcFile.putAttrInt( mesh2FaceNodesId, "_FillValue", -999 );
+
+  // Turning off define mode - allows data write
+  nc_enddef( mNcFile.handle() );
+
+  // Write vertices
+  std::unique_ptr<MDAL::MeshVertexIterator> vertexIterator = mesh->readVertices();
+
+  const size_t verticesCoordCount = mesh->verticesCount() * 3;
+  double *verticesCoordinates = new double[verticesCoordCount];
+
+  vertexIterator->next( mesh->verticesCount(), verticesCoordinates );
+
+  for ( size_t i = 0; i < mesh->verticesCount(); i++ )
+  {
+    mNcFile.putDataDouble( mesh2dNodeXId, i, verticesCoordinates[3 * i] );
+    mNcFile.putDataDouble( mesh2dNodeYId, i, verticesCoordinates[3 * i + 1] );
+    mNcFile.putDataDouble( mesh2dNodeZId, i, verticesCoordinates[3 * i + 2] );
+  }
+
+  delete[] verticesCoordinates;
+
+  // Write faces
+  std::unique_ptr<MDAL::MeshFaceIterator> faceIterator = mesh->readFaces();
+  const size_t faceVerticesMax = mesh->faceVerticesMaximumCount();
+
+  int faceOffsets[1];
+  int *vertexIndices = new int[faceVerticesMax];
```

Comment:
do not use raw vectors. Also do not iterate one by one, it is not speed-effective. see https://github.com/qgis/QGIS/blob/b2e712149312390b483cfb3c9be4c63e875694f0/src/providers/mdal/qgsmdalprovider.cpp#L116 how to iterate in chunks